File: server/pet/identity.go
package pet

import (
	"errors"
	"fmt"
	"regexp"

	"cloud-clicker/server/decimal"
)

// Identity is one adopted pet's immutable identity (Pet Adoption v1 PA3). It
// is keyed by pet ID in Founder save pet_identities, beside the mutable care
// record in pets.
type Identity struct {
	SpeciesID           string `json:"species_id"`
	Temperament         string `json:"temperament"`
	PaletteID           string `json:"palette_id"`
	NameKey             string `json:"name_key"`
	AdoptedAtMS         int64  `json:"adopted_at_ms"`
	AdoptedAtAttendedMS int64  `json:"adopted_at_attended_ms"`
}

var (
	ErrInvalidIdentity = errors.New("invalid pet identity")
	petIDV7Pattern     = regexp.MustCompile(`^[0-9a-f]{8}-[0-9a-f]{4}-7[0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12}$`)
)

// ValidateIdentityShape checks the catalog-free PA3 invariants: exact
// biconditional key sets with care, UUIDv7 IDs, well-formed fields, and an
// adoption coordinate no later than the care watermark.
func ValidateIdentityShape(identities map[string]Identity, care map[string]CareState) error {
	if identities == nil || len(identities) != len(care) {
		return fmt.Errorf("%w: identity and care key sets differ", ErrInvalidIdentity)
	}
	for id, identity := range identities {
		record, ok := care[id]
		if !ok || !petIDV7Pattern.MatchString(id) {
			return fmt.Errorf("%w: pet %q has no care record or is not UUIDv7", ErrInvalidIdentity, id)
		}
		if !speciesIDPattern.MatchString(identity.SpeciesID) || !contains(TemperamentOrder, identity.Temperament) ||
			!paletteIDPattern.MatchString(identity.PaletteID) || !nameKeyPattern.MatchString(identity.NameKey) {
			return fmt.Errorf("%w: pet %q has a malformed identity", ErrInvalidIdentity, id)
		}
		if identity.AdoptedAtMS < 0 || identity.AdoptedAtMS > decimal.MaxExactInteger || identity.AdoptedAtAttendedMS < 0 ||
			identity.AdoptedAtAttendedMS > record.EvaluatedThroughAttendedMS {
			return fmt.Errorf("%w: pet %q has an invalid adoption coordinate", ErrInvalidIdentity, id)
		}
	}
	return nil
}

// ValidateIdentitiesAgainst adds the catalog-aware PA3.3/PA3.4 invariants.
func ValidateIdentitiesAgainst(catalog *SpeciesCatalog, identities map[string]Identity) error {
	if catalog == nil {
		return fmt.Errorf("%w: pet_species is not pinned", ErrInvalidIdentity)
	}
	if int64(len(identities)) > catalog.MaxPetsPerFounder {
		return fmt.Errorf("%w: %d pets exceed the cap %d", ErrInvalidIdentity, len(identities), catalog.MaxPetsPerFounder)
	}
	for id, identity := range identities {
		row, ok := catalog.Row(identity.SpeciesID)
		if !ok || !row.HasTemperament(identity.Temperament) || !row.HasPalette(identity.PaletteID) || !row.HasName(identity.NameKey) {
			return fmt.Errorf("%w: pet %q does not resolve under the pinned pet_species", ErrInvalidIdentity, id)
		}
	}
	return nil
}

func CloneIdentities(values map[string]Identity) map[string]Identity {
	if values == nil {
		return nil
	}
	result := make(map[string]Identity, len(values))
	for id, identity := range values {
		result[id] = identity
	}
	return result
}

// InitialCareState is PA4.6.2's canonical initial care record for a pet
// adopted at the frozen attendance total attendedMS: stats at their initial
// values, zero remainders, no cooldowns, Trust at its initial value, and a
// watermark at attendedMS so pre-adoption attendance never decays the pet.
func InitialCareState(catalog *Catalog, attendedMS int64) (CareState, error) {
	if catalog == nil || attendedMS < 0 || attendedMS > decimal.MaxExactInteger {
		return CareState{}, fmt.Errorf("%w: invalid initial care inputs", ErrInvalidIdentity)
	}
	stats, remainders := map[StatID]int64{}, map[StatID]int64{}
	for _, row := range catalog.StatPolicy.Stats {
		stats[row.StatID], remainders[row.StatID] = row.InitialPPM, 0
	}
	return CareState{StatsPPM: stats, StatDecayRemaindersPPM: remainders, CooldownUntilAttendedMS: map[string]int64{},
		TrustPPM: catalog.TrustPolicy.InitialPPM, EvaluatedThroughAttendedMS: attendedMS, BehaviorState: BehaviorIdle,
		BehaviorEnteredAtAttendedMS: attendedMS, BehaviorQueue: []BehaviorQueueEntry{}}, nil
}

File: client/src/pet/identity.ts
import { MAX_EXACT_INTEGER } from "../numeric";
import type { PetCatalog } from "./catalog";
import type { PetStatID } from "./grammar";
import { PET_TEMPERAMENT_ORDER, petSpeciesRow, type PetSpeciesCatalog, type PetTemperament } from "./species";
import type { PetCareState } from "./state";

// Pet Adoption v1 PA3: the immutable identity of an adopted pet, keyed like
// the care map. Byte twin of server/pet/identity.go.
export interface PetIdentity {
  readonly species_id: string;
  readonly temperament: PetTemperament;
  readonly palette_id: string;
  readonly name_key: string;
  readonly adopted_at_ms: number;
  readonly adopted_at_attended_ms: number;
}

const identityKeys = ["adopted_at_attended_ms", "adopted_at_ms", "name_key", "palette_id", "species_id", "temperament"] as const;
const petIDV7 = /^[0-9a-f]{8}-[0-9a-f]{4}-7[0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12}$/u;
const speciesID = /^pet_species\.[a-z][a-z0-9_]*$/u;
const paletteID = /^pet_palette\.[a-z0-9_]+$/u;
const nameKey = /^pet\.name\.[a-z0-9_]+\.[a-z0-9_]+$/u;

export function parsePetIdentities(source: unknown, care: Readonly<Record<string, PetCareState>>, species: PetSpeciesCatalog): Record<string, PetIdentity> {
  if (source === null || typeof source !== "object" || Array.isArray(source)) throw new SyntaxError("pet_identities must be an object");
  const result: Record<string, PetIdentity> = {};
  for (const [id, raw] of Object.entries(source as Record<string, unknown>)) {
    if (raw === null || typeof raw !== "object" || Array.isArray(raw)) throw new SyntaxError("pet identity must be an object");
    const keys = Object.keys(raw).sort();
    if (keys.length !== identityKeys.length || keys.some((key, index) => key !== identityKeys[index])) throw new SyntaxError("pet identity keys are not exact");
    const row = raw as Record<string, unknown>;
    result[id] = Object.freeze({ species_id: row.species_id as string, temperament: row.temperament as PetTemperament, palette_id: row.palette_id as string,
      name_key: row.name_key as string, adopted_at_ms: row.adopted_at_ms as number, adopted_at_attended_ms: row.adopted_at_attended_ms as number });
  }
  validatePetIdentities(result, care, species);
  return result;
}

export function validatePetIdentities(identities: Readonly<Record<string, PetIdentity>>, care: Readonly<Record<string, PetCareState>>, species: PetSpeciesCatalog): void {
  const ids = Object.keys(identities);
  if (ids.length !== Object.keys(care).length) throw new SyntaxError("pet identity and care key sets differ");
  if (ids.length > species.max_pets_per_founder) throw new SyntaxError("pet identities exceed the cap");
  for (const id of ids) {
    const identity = identities[id]!, record = care[id];
    if (!record || !petIDV7.test(id)) throw new SyntaxError("pet identity has no care record or is not UUIDv7");
    if (typeof identity.species_id !== "string" || !speciesID.test(identity.species_id) || !PET_TEMPERAMENT_ORDER.includes(identity.temperament) ||
      typeof identity.palette_id !== "string" || !paletteID.test(identity.palette_id) || typeof identity.name_key !== "string" || !nameKey.test(identity.name_key)) {
      throw new SyntaxError("pet identity is malformed");
    }
    if (!Number.isSafeInteger(identity.adopted_at_ms) || identity.adopted_at_ms < 0 || identity.adopted_at_ms > MAX_EXACT_INTEGER ||
      !Number.isSafeInteger(identity.adopted_at_attended_ms) || identity.adopted_at_attended_ms < 0 || identity.adopted_at_attended_ms > record.evaluated_through_attended_ms) {
      throw new SyntaxError("pet identity has an invalid adoption coordinate");
    }
    const row = petSpeciesRow(species, identity.species_id);
    if (!row || !row.allowed_temperaments.includes(identity.temperament) || !row.palette_ids.includes(identity.palette_id) || !row.name_keys.includes(identity.name_key)) {
      throw new SyntaxError("pet identity does not resolve under the pinned pet_species");
    }
  }
}

export function encodePetIdentities(identities: Readonly<Record<string, PetIdentity>>): Record<string, PetIdentity> {
  return Object.fromEntries(Object.keys(identities).sort().map((id) => {
    const identity = identities[id]!;
    return [id, { adopted_at_attended_ms: identity.adopted_at_attended_ms, adopted_at_ms: identity.adopted_at_ms, name_key: identity.name_key,
      palette_id: identity.palette_id, species_id: identity.species_id, temperament: identity.temperament }];
  }));
}

// PA4.6.2's canonical initial care record, watermarked at the frozen
// attendance total so pre-adoption attendance never decays the pet.
export function initialPetCareState(catalog: PetCatalog, attendedMs: number): PetCareState {
  if (!Number.isSafeInteger(attendedMs) || attendedMs < 0 || attendedMs > MAX_EXACT_INTEGER) throw new RangeError("invalid initial care attendance");
  const stats = Object.fromEntries(catalog.stat_policy.stats.map((row) => [row.stat_id, row.initial_ppm])) as Record<PetStatID, number>;
  const remainders = Object.fromEntries(catalog.stat_policy.stats.map((row) => [row.stat_id, 0])) as Record<PetStatID, number>;
  return { stats_ppm: stats, stat_decay_remainders_ppm: remainders, cooldown_until_attended_ms: {}, trust_ppm: catalog.trust_policy.initial_ppm,
    trust_decay_remainder_ppm: 0, evaluated_through_attended_ms: attendedMs, behavior_state: "idle", behavior_entered_at_attended_ms: attendedMs,
    behavior_queue: [], behavior_prng_cursor: 0 };
}
